core/lib/scmi-server/scmi_server_scpfw.c

```diff
@@ -0,0 +1,329 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2024, STMicroelectronics International N.V.
+ */
+
+#include <assert.h>
+#include <drivers/scmi.h>
+#include <kernel/boot.h>
+#include <kernel/dt.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <scmi_agent_configuration.h>
+#include <stdlib.h>
+#include <sys/queue.h>
+#include <util.h>
+
+/*
+ * struct optee_scmi_server - Data of scmi_server_scpfw
+ *
+ * @dt_name: SCMI node name
+ * @agent_list: list of optee_scmi_server_agent
+ */
+struct optee_scmi_server {
+	const char *dt_name;
+	SIMPLEQ_HEAD(, optee_scmi_server_agent) agent_list;
+};
+
+/*
+ * @struct optee_scmi_server_agent - Data of an SCMI agent
+ *
+ * @dt_name: SCMI agent node name
+ * @agent_id: SCMI agent identifier
+ * @channel_id: SCMI channel identifier
+ * @protocol_list: list of optee_scmi_server_protocol
+ * @link: link for optee_scmi_server:agent_list
+ */
+struct optee_scmi_server_agent {
+	const char *dt_name;
+	unsigned int agent_id;
+	unsigned int channel_id;
+	SIMPLEQ_HEAD(, optee_scmi_server_protocol) protocol_list;
+	SIMPLEQ_ENTRY(optee_scmi_server_agent) link;
+};
+
+/*
+ * struct optee_scmi_server_protocol - Data of an SCMI protocol
+ *
+ * @dt_name: SCMI protocol node name
+ * @dt_node: SCMI protocol node
+ * @protocol_id: SCMI protocol identifier
+ * @link: list for optee_scmi_server_agent:protocol_list
+ */
+struct optee_scmi_server_protocol {
+	const char *dt_name;
+	int dt_node;
+	unsigned int protocol_id;
+	SIMPLEQ_ENTRY(optee_scmi_server_protocol) link;
+};
+
+/* scmi_agent_configuration API */
+static struct scpfw_config scpfw_cfg;
+
+static void scmi_scpfw_free_agent(struct scpfw_agent_config *agent_cfg)
+{
+	free(agent_cfg->channel_config);
+}
+
+struct scpfw_config *scmi_scpfw_get_configuration(void)
+{
+	struct scpfw_agent_config *old_agent_config = scpfw_cfg.agent_config;
+
+	assert(scpfw_cfg.agent_count >= 1);
+	assert(!old_agent_config[0].channel_count);
+
+	/*
+	 * The agents config data passed to SCP-firmware do not consider
+	 * agent ID 0 that is reserved for the SCMI server itself in
+	 * SCP-firmware and therefore has no configuration data.
+	 */
+	scpfw_cfg.agent_count--;
+	scpfw_cfg.agent_config = calloc(scpfw_cfg.agent_count,
+					sizeof(*scpfw_cfg.agent_config));
+
+	memcpy(scpfw_cfg.agent_config, old_agent_config + 1,
+	       sizeof(*scpfw_cfg.agent_config) * scpfw_cfg.agent_count);
```

Comment:
Maybe it would be worth preventing this `memcpy()` and change your SCP-fmw M-R to consider agent #0 cell description as being the SCMI server/platform description (that is it should have no channel described).

Reply:
SCP-firmware merge request ([M-R 1143](https://gitlab.arm.com/firmware/SCP-firmware/-/merge_requests/1143)) seems happy with the current implementation so I think you can discard my comment and keep this as-is.